### PR TITLE
fix: catalog pagination

### LINKED ISSUE #2493: Offers catalog pagination not working as expected
# Bug Report

## Describe the Bug
This bug relates to #2008.
The fix applied in #2018 didn't resolve the issue completely.
When requesting the catalog using `limit` and `offset` queryparams, the number of contractOffers returned does not correspond to the expected.

## Steps to Reproduce
Test scenario:
 - 2 Contract definitions with asset selector criteria refering to 50 assets each (Expected total of 100 offers).
 - Consumer requests the catalog from provider using different `limit` and `offset`.

### Expected Behavior
**Behaviour a)**
limit=52
offset=0
expected offers: 52

**Behaviour b)**
limit=2
offset=49
expected offers: 2

### Observed Behavior
**Behaviour a)**
observed offers: 100

**behaviour b)**
observed offers: 3

### Used version
0.0.1-20230213-SNAPSHOT

## Possible Implementation
On `ContractOfferResolverImpl.queryContractOffers()` when `skip` is a value under 0, `limit` for next iteration should be the result of the current `limit` minus the `skip`.

### PR DESCRIPTION
**Pull request by yurimssilva**
## What this PR changes/adds

- Adds test cases to evaluate catalog pagination across multiple contract definitions.
- Changes the logic which dynamically builds the offset and limit for each contract definition.

## Why it does that

  Previously the number of assets that should be skipped in the current iteration was dynamic but the limit was not, which caused the pagination to fail when across multiple contract definitions.
  I decided to use a different approach that for me was easier to understand.
  Based on the given range, the absolute offset and limit are static but could assume a different state on each iteration responding to the number of seen and fetched assets until that point. Therefore, I implemented the logic that sets the values of the dynamic offset and limit variables in the following expressions:

- The dynamic offset must be equal to the absolute offset minus the number of assets seen, but never less than zero.
- The dynamic limit must be equal to the absolute limit minus the number of assets already fetched, but never exceed the number of assets obtained in the current contract definition.

## Linked Issue(s)

Closes #2493

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [x] assigned appropriate label? (exclude from changelog with label `no-changelog`)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-edc/Connector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [Etiquette for pull requests](https://github.com/eclipse-edc/Connector/blob/main/pr_etiquette.md) for details_)

### CODE REVIEW

File: core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java [hunk 1]
```diff
@@ -107,6 +102,35 @@ void shouldLimitResult_withHeterogenousChunks() {
...
+
+        var from = 0;
+        var to = 12;
+        var query = ContractOfferQuery.builder().range(new Range(from, to)).claimToken(ClaimToken.Builder.newInstance().build()).build();
+        assertThat(contractOfferResolver.queryContractOffers(query)).hasSize(12);
+
+        from = 8;
+        to = 14;
+        query = ContractOfferQuery.builder().range(new Range(from, to)).claimToken(ClaimToken.Builder.newInstance().build()).build();
+        assertThat(contractOfferResolver.queryContractOffers(query)).hasSize(6);
```
Comment (paullatzelsperger) [inline core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java:136]: nit: may want to use variables here, i.e. `to - from`

File: core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java [hunk 2]
```diff
@@ -107,6 +107,35 @@ void shouldLimitResult_withHeterogenousChunks() {
...
+
+        var contractDefinition1 = getContractDefBuilder("contract-definition-")
+                .selectorExpression(selectorFrom(assets1)).build();
+        var contractDefinition2 = getContractDefBuilder("contract-definition-")
+                .selectorExpression(selectorFrom(assets2)).build();
+
+        when(agentService.createFor(isA(ClaimToken.class))).thenReturn(new ParticipantAgent(emptyMap(), emptyMap()));
+        when(contractDefinitionService.definitionsFor(isA(ParticipantAgent.class))).thenAnswer(i -> Stream.of(contractDefinition1, contractDefinition2));
+        when(policyStore.findById(any())).thenReturn(PolicyDefinition.Builder.newInstance().policy(Policy.Builder.newInstance().build()).build());
+
```
Comment (paullatzelsperger) [inline core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java:127]: those two assertions seem to test the same thing, i.e. getting contract offers that "span" across assets that are referred to by two contract definitions. Therefor one of them could be removed. However, we should actually test edge cases too:
- `0...12`
- `0...9999`: tests a range overshoot, expect 20 offers
- `4...9999`: test a range overshoot starting at an offset, expect 16 offers
- `3...20`: test exact upper bound with offset, expect 17 offers

At that point you may want to consider a `@ParameterizedTest` to avoid test code duplication.

**Review by ndr-brt — APPROVED** [review]

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
Thanks for the contribution @yurimssilva. Please try to improve your PR description so that people can understand _what_ exactly you changed:
> Changes the logic which dynamically builds the offset and limit for each contract definition

yeah, but _how_ did you change the logic? what was the bug you fixed? Reviewers should not be expected to figure this out on their own. Some more comments inline.

Also Unit tests and checkstyle are failing.

File: core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java [hunk 3]
```diff
@@ -50,12 +50,7 @@
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.isA;
-import static org.mockito.Mockito.atLeastOnce;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.never;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
[the remark this answers is not in the document]
Comment (yurimssilva) [reply core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java]: ty, fixed

File: core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java (hunk 1 above)
Comment (yurimssilva) [reply core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java:136]: ty, fixed.

File: core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java (hunk 2 above)
Comment (yurimssilva) [reply core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java:127]: thanks, @paullatzelsperger.. doing it.

File: core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java [hunk 4]
```diff
@@ -107,6 +112,42 @@ void shouldLimitResult_withHeterogenousChunks() {
...
+                Arguments.of(8, 14),
+                Arguments.of(0, 999),
+                Arguments.of(4, 888),
+                Arguments.of(3, 20),
+                Arguments.of(23, 25)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("ranges")
```
Comment (paullatzelsperger) [inline core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java:127]: I'd rather you use `@CsvSource`, because then at least the test values are immediately visible, or `@ArgumentSource`, because that uses a strongly typed provider rather thant relying solely on the method name.
Also, the method naming should follow the `should_when_then` pattern, separated by an underscore, which increases legibility quite a bit.

File: core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java (hunk 2 above)
Comment (paullatzelsperger) [reply core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java:146]: we're getting there :) check updated comments.

File: core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java (hunk 4 above)
Comment (yurimssilva) [reply core/control-plane/contract-core/src/test/java/org/eclipse/edc/connector/contract/offer/ContractOfferResolverImplIntegrationTest.java]: thx, Paul.. it's fixed.

**Review by paullatzelsperger — APPROVED** [review]
